### DISCUSSION TITLE
[4.0.9] HTTP2 not activating?

### DISCUSSION THREAD

**Discussion by hrstoyanov** [helidon-io/helidon#8828] [Q&A] [answered]
For my web app, I made sure this dependency :
```
io.helidon.webclient:helidon-webclient-http2:4.0.9
```

is included in my class path. Therefore, [according to the 4.x docs](https://helidon.io/docs/v4/se/webserver#_http2_support) this is enough to enable/trigger/upgrade to  **http2** protocol.
However, when I start my web app and look a at Chrome Dev tools Network tab , the "Protocol" columns says **http/1.1.**

**Why? How can I enforce HTTP2 (maybe programmatically?) even for local testing with http://localhost:8080?**  Could not find anything in the examples...
 
Thanks.

<img width="809" alt="Screenshot 2024-06-02 at 4 55 25 AM" src="https://github.com/helidon-io/helidon/assets/5327044/c160bb1c-e9db-4eb4-9964-0b6b409b24b5">

**Comment by tomas-langer** [comment]
Support for HTTP/2 on the server is:
```xml
<dependency>
      <groupId>io.helidon.webserver</groupId>
      <artifactId>helidon-webserver-http2</artifactId>
</dependency>
```

To verify that it works, use a client that supports HTTP/2, such as `curl`:
```
curl -i --http2 http://localhost:8080/simple-greet
```

and you get similar answer (if it works):
```
HTTP/1.1 101 Switching Protocols
Connection: Upgrade
Upgrade: h2c

HTTP/2 200
content-length: 12
content-type: text/plain; charset=UTF-8
date: Mon, 3 Jun 2024 16:41:28 +0200

Hello World!%  
```

(this is output based on Helidon 4.0.9 quickstart with added HTTP/2 server library)

**Comment by tomas-langer** [comment]
When using TLS (`https`), ALPN (Application layer protocol negotiation) is used, and the client agrees with the server on a supported HTTP version - this is supported by Helidon WebServer and WebClient.

**Reply by hrstoyanov** [reply]
Thanks @tomas-langer ,
But the only client I need to work with http2 at the moment is Chrome (or any modern) browser, not curl or web client app . 

As you can see from the screenshot, out of the box, the browser never switches into http2, and therefore all .CSS and other web assets are not bundled .

I expected after the initial GET, helidon to propose an upgrade the connection to http2, so all subsequent GETs from the browser follow over http2. 

Any idea why this is not happening and all GETs remain http1.1?

**Reply by hrstoyanov** [reply, marked as answer]
@tomas-langer I might have found the answer to my own question: Browsers do not support http2 over unencrypted connection, so I need to use https, certificates, etc.  even for localhost testing! Ugh...

**Reply by tomas-langer** [reply]
When https is enabled, the following can be seen:
![Screenshot from 2024-06-04 17-33-03](https://github.com/helidon-io/helidon/assets/13766491/9cf9bccc-f12c-43e1-b027-a845346b585c)

**Reply by tomas-langer** [reply]
This is the configuration I have used (using PKCS-12 keystore):
```yaml
server:
  port: 8080
  host: 0.0.0.0
  tls:
    private-key:
      keystore:
        passphrase: "changeit"
        resource:
          resource-path: "server.p12"
```

**Comment by hrstoyanov** [comment]
For anyone needing to test http2 via Chrome in a  local dev machine without the hassle:
1. Use the awesome [mkcert](https://github.com/FiloSottile/mkcert) to generate certs and deploy them to your entire local machine
2. The generated certs will be available for your browsers, since all of them consult the machine's cert store 
3. To make Helidon (or any Java app) also see trust the system certs, start it by overriding the jvm system trust store  option:
 - for windows:  -Djavax.net.ssl.trustStoreType=WINDOWS-ROOT
 - for macos: -Djavax.net.ssl.trustStoreType=KeychainStore 
 - (or if you want to use both ones: -Djavax.net.ssl.trustStoreType=KeychainStore -Djavax.net.ssl.trustStore=/Library/Keychains/System.keychain)
